### PR TITLE
refactor with ibis

### PR DESCRIPTION
**Pull request by mayurkmmt**
### **User description**
feat: refactor get_index function with ibis

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Refactored `get_indexes` function to use Ibis for scalable computation.

- Enhanced `get_indexes` to support both pandas DataFrame and Ibis Table.

- Added robust error handling for invalid filters and aggregation functions.

- Updated and expanded test cases for `get_indexes` with various scenarios.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>index.py</strong><dd><code>Refactor `get_indexes` to use Ibis and enhance functionality</code></dd></summary>
<hr>

pyretailscience/plots/index.py

<li>Refactored <code>get_indexes</code> to use Ibis for efficient computation.<br> <li> Added support for both pandas DataFrame and Ibis Table.<br> <li> Improved error handling for invalid filters and unsupported <br>aggregation functions.<br> <li> Updated logic for calculating proportions and indices.

</details>

  </td>
  <td><a href="https://github.com/Data-Simply/pyretailscience/pull/95/files#diff-5e354290215a1b5f7226d508e6766857d97672b3d53f2ddad9dd4b508625f7ee">+45/-14</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_index.py</strong><dd><code>Update and expand test cases for `get_indexes`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/plots/test_index.py

<li>Replaced outdated test cases with new ones for <code>get_indexes</code>.<br> <li> Added tests for invalid filters and unsupported aggregation functions.<br> <li> Included tests for various aggregation functions and offset handling.<br> <li> Simplified and clarified test data and assertions.

</details>

  </td>
  <td><a href="https://github.com/Data-Simply/pyretailscience/pull/95/files#diff-7aa80c2a38d9bbd037f8171a4b832aaaffc9f55cbd318c04de31b02176689474">+67/-85</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

## Summary by CodeRabbit

- **New Features**
  - Updated plotting functionality with new parameters for enhanced index calculation.
  - Improved versatility of index retrieval by supporting both DataFrame and Ibis Table inputs.

- **Refactor**
  - Enhanced data processing for faster and more reliable visualizations, improving performance on larger datasets.
  - Updated aggregation logic now delivers a more comprehensive view of grouped data.

- **Tests**
  - Expanded test coverage to ensure robust handling of various input conditions and edge cases, including validation for invalid aggregation functions and improved clarity in test cases.

### CODE REVIEW

File: pyretailscience/plots/index.py [hunk 1]
```diff
@@ -239,18 +240,18 @@ def plot(  # noqa: C901, PLR0913 (ignore complexity and line length)
 
 
 def get_indexes(
-    df: pd.DataFrame,
+    df: pd.DataFrame | ibis.Table,
     df_index_filter: list[bool],
```
Comment (mvanwyk) [inline pyretailscience/plots/index.py:244]: Hey @mayurkmmt can you remove the concept of the index. See the original ticket. This index concept won't work at scale. You might have to pass a list of a billion rows or more. My thinking is to split it into parameters. An index_col and value_to_index., parameters. Eg the below. Let me know if you think the naming is confusing.
```python
from pyretailscience.standard_graphs import index_plot

index_plot(
    df,
    index_col="segment_name",
    value_to_index="Light",
    value_col="unit_price",
    group_col="category_0_name",
)
```

File: tests/plots/test_index.py
Comment (mvanwyk) [inline tests/plots/test_index.py:1]: I don't see any tests that the index values are calculated correctly. They were there before. Can you add those back please.

File: pyretailscience/plots/index.py [hunk 2]
```diff
@@ -239,45 +242,75 @@ def plot(  # noqa: C901, PLR0913 (ignore complexity and line length)
...
-        overall_total = overall_df.groupby(index_subgroup_col)[value_col].sum()
-    overall_s = overall_df[value_col] / overall_total
+        overall_total = overall_agg.group_by(index_subgroup_col).aggregate(total=lambda t: t.value.sum())
+        overall_props = (
+            overall_agg.join(overall_total, index_subgroup_col)
+            .mutate(proportion=lambda t: t.value / t.total)
+            .drop("total")
+        )
+
+    overall_props = overall_props.mutate(proportion_overall=overall_props.proportion).drop("proportion")
```
[the remark this answers is not in the document]
Comment (mvanwyk) [reply pyretailscience/plots/index.py:296]: Can we just name it proportion_overall in the first place then and remove this line.

File: tests/plots/test_index.py [hunk 3]
```diff
@@ -8,126 +8,180 @@
...
+        value_col="value",
+        group_col="category",
+        offset=5,
+    )
+
+    assert isinstance(result, pd.DataFrame)
+    assert "category" in result.columns
+    assert "index" in result.columns
+    assert not result.empty
+    assert all(result["index"] >= OFFSET_THRESHOLD)
```
Comment (mvanwyk) [inline tests/plots/test_index.py:119]: Since OFFSET_THRESHOLD is only used in this function, can you move its instantiation here please.

File: tests/plots/test_index.py [hunk 4]
```diff
@@ -8,126 +8,180 @@
...
+            "value": [10, 20, 30],
+        },
+    )
+    result = get_indexes(
+        df,
+        value_to_index="A",
+        index_col="category",
+        value_col="value",
+        group_col="category",
+        offset=5,
```
Comment (mvanwyk) [inline tests/plots/test_index.py:112]: SInce this is tied to the value of OFFSET_THRESHOLD I think we should change this line.
```suggestion
        offset=-OFFSET_THRESHOLD,
```

File: tests/plots/test_index.py [hunk 5]
```diff
@@ -8,126 +8,180 @@
...
+        {
+            "group_col": ["A"],
+            "value": [3],
+            "proportion": [1.0],
+            "value_right": [3],
+            "proportion_overall": [0.142857],
+            "index": [700.0],
+        },
+    )
+
```
Comment (mvanwyk) [inline tests/plots/test_index.py:141]: This isn't right. See the previous tests. There should be a value create for each of the values of group_col. ie A,B 
 and C. Also, the other columns should be removed. Only group_col and index should be there.
 
 https://github.com/Data-Simply/pyretailscience/blob/main/tests/plots/test_index.py#L22

File: tests/plots/test_index.py [hunk 6]
```diff
@@ -8,126 +8,180 @@
...
-            "index": [140, 140, 46.6666667],
+            "group_col1": ["A", "A"],
+            "group_col2": ["D", "E"],
+            "value": [3, 15],
+            "proportion": [0.166667, 0.833333],
+            "value_right": [3, 15],
+            "proportion_overall": [0.166667, 0.833333],
+            "index": [100.0, 100.0],
         },
     )
```
Comment (mvanwyk) [inline tests/plots/test_index.py:172]: This isn't also right. See the previous tests. There should be a value created for each of the values of group_col2 and group_col2. Also, the other columns should be removed. Only group_col1, group_col2 and index should be there.
 
https://github.com/Data-Simply/pyretailscience/blob/main/tests/plots/test_index.py#L42

**Review by mvanwyk — APPROVED** [review]
LGTM!
